tests/regression/src/lib.rs

```diff
@@ -104,32 +104,158 @@ mod tests {
+        Valgrind,
+        Diff,
     }
 
-    /// Function to determine if diff mode is enabled, runs diff assertion if it is.
-    fn is_diff_mode() -> bool {
-        env::var("DIFF_MODE").is_ok()
+    impl RegressionTestMode {
+        fn from_env() -> Self {
+            if env::var("ENABLE_VALGRIND").is_ok() {
+                RegressionTestMode::Valgrind
+            } else if env::var("DIFF_MODE").is_ok() {
+                RegressionTestMode::Diff
+            } else {
+                RegressionTestMode::Default
+            }
+        }
     }
 
     fn valgrind_test<F>(test_name: &str, test_body: F) -> Result<(), s2n_tls::error::Error>
     where
         F: FnOnce(&InstrumentationControl) -> Result<(), s2n_tls::error::Error>,
     {
-        if !is_running_under_valgrind() {
-            if is_diff_mode() {
+        match RegressionTestMode::from_env() {
+            RegressionTestMode::Valgrind => {
+                run_valgrind_test(test_name);
+                Ok(())
+            },
+            RegressionTestMode::Diff => {
                 assert_performance_diff(test_name);
                 Ok(())
-            } else {
+            },
+            RegressionTestMode::Default => {
                 let ctrl = InstrumentationControl;
                 test_body(&ctrl)
+            },
+        }
+    }
+
+    struct RawProfile {
+        test_name: String,
+        commit_hash: String,
+    }
+
+    impl RawProfile {
+        fn new(test_name: &str, commit_hash: &str) -> Self {
+            let new_dir = format!("target/{}", commit_hash);
+            create_dir_all(Path::new(&new_dir)).unwrap();
+            Self {
+                test_name: test_name.to_string(),
+                commit_hash: commit_hash.to_string(),
+            }
+        }
+
+        fn path(&self) -> String {
+            format!("target/{}/{}.raw", self.commit_hash, self.test_name)
+        }
+
+        fn annotate(&self) -> AnnotatedProfile {
+            let output_file = self.path();
+            let annotate_output = Command::new("cg_annotate")
+                .arg(&output_file)
+                .output()
+                .expect("Failed to run cg_annotate");
+            if !annotate_output.status.success() {
+                panic!("cg_annotate failed");
+            }
+            let annotate_content = String::from_utf8(annotate_output.stdout).expect("Invalid UTF-8 in cg_annotate output");
+            let annotated_path = format!("target/{}/{}.annotated", self.commit_hash, self.test_name);
+            write_to_file(&annotated_path, &annotate_content);
+            AnnotatedProfile {
+                path: annotated_path,
+                content: annotate_content,
+            }
+        }
+
+        fn execute_command(&self, mut command: Command) {
+            let output = command.output().expect("Failed to execute command");
+            if !output.status.success() {
+                panic!("Command failed: {:?}", output);
+            }
```

Comment:
```suggestion
            assert!(output.status.success());
```
Since we have to do this pretty frequently (and this is just a test) it might be neater to assert like this.